tests/mode_test.py

```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/python3
+
+# Test that we can read the sensor modes, and then configure the camera
+# into each mode and get the correct framerate
+
+from picamera2 import Picamera2
+import time
+from math import isclose
+
+picam2 = Picamera2()
+
+for mode in picam2.sensor_modes:
+    picam2.video_configuration = picam2.create_video_configuration(
+        raw={"size": mode["size"], "format": mode["format"].format},
+    )
+    picam2.configure("video")
+    picam2.set_controls({"FrameRate": int(mode["fps"])})
+    picam2.start()
+    # Check we got roughly the right framerate
+    print(f"Framerate: {picam2.controls.FrameRate:.2f}, requested: {mode['fps']:.2f}")
+    assert isclose(picam2.controls.FrameRate, round(mode["fps"]), abs_tol=1.0)
```

Comment:
I'm not sure checking this really adds very much... as you'll have seen in controls.py, you're really just retrieving the same framerate that you set earlier!

Rather than recording a video, one idea might be just to run the camera for, say, 1 second and then capture the metadata. This should contain a "FrameDuration", which is actually measured from frames delivered by the camera, and should tie up with the number you requested. Of course, it requires quite a lot of stuff to be working correctly, right down to the register and timing calculations in the sensor driver!